File: com/generic/sync/CountDownLatch.java
package com.generic.sync;

public class CountDownLatch {
    private int count;

    public CountDownLatch(int nThreads) {
        this.count = nThreads;
    }

    public synchronized void countDown() {
        count--;
        if (count == 0) {
            synchronized (this) {
                this.notifyAll();
            }
        } else {
            try {
                synchronized (this) {
                    this.wait();
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public int getCount() {
        return this.count;
    }
}

File: csharp_common/csharp_common/CountDownLatch.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

namespace csharp_common
{
    class CountDownLatch
    {
        private int count { get; set; }

        public CountDownLatch(int nThreads)
        {
            this.count = nThreads;
        }

        public void countDown()
        {
            count--;
            if (count == 0)
            {
                lock (this) { Monitor.PulseAll(this); }
            }
            else
            {
                lock(this) { Monitor.Wait(this); }
            }
        }
    }
}
